src/main.ts

```diff
@@ -89,10 +90,38 @@ function setupTrayItem() {
 
 function handleDownloads() {
     mainWindow.webContents.session.on('will-download', (event, item) => {
-        const savePath = path.join(
-            app.getPath('downloads'),
-            item.getFilename()
+        item.setSavePath(
+            getUniqueSavePath(item.getFilename(), app.getPath('downloads'))
         );
-        item.setSavePath(savePath);
     });
 }
+
+function getUniqueSavePath(filename: string, directory: string): string {
+    let n = 0;
+    let exists;
+    let uniqueFileSavePath;
+    const [filenameWithoutExtension, extension] =
+        splitFilenameAndExtension(filename);
+    do {
+        let fileNameWithNumberedSuffix;
+        if (n > 0) {
+            fileNameWithNumberedSuffix = `${filenameWithoutExtension}(${n}).${extension}`;
+        } else {
+            fileNameWithNumberedSuffix = filename;
+        }
+        uniqueFileSavePath = path.join(directory, fileNameWithNumberedSuffix);
+        exists = existsSync(uniqueFileSavePath);
+        n++;
+    } while (exists);
+    return uniqueFileSavePath;
+}
+
+function splitFilenameAndExtension(filename: string): [string, string] {
+    const lastDotPosition = filename.lastIndexOf('.');
+    if (lastDotPosition === -1) return [filename, null];
```

Comment:
This is returning `null` even though the type is `string` - why isn't tsc complaining?

This would cause a bug if there is a filename without an extension - when we do string interpolation later we'd add 'null'.

Also, this method shouldn't be needed? There is already a split in JS

```
const [nameWithoutExtension, extension] = 'testing.png'.split('.')
```

In case the extension is missing, `extension` would be `undefined`. But we don't have to add checks for that, we can safely reconstruct the name by using the reverse - `join` - it'll work fine even if extension is not there.

```
[`nameWithoutExtension-${n}`, extension].join('')
```

Reply:
> This is returning null even though the type is string - why isn't tsc complaining?

we have strict mode disabled 😅 .

😮 ohh sweet... Ya, using split and join is better... didn't think it through... have done it on web earlier, just copy pasted the code from there 😬 , will fix there too